### PR TITLE
fix: The  aggs result of NestedAggregator with sub NestedAggregator may be not accurately

### LINKED ISSUE #13303: [BUG] The aggs result of NestedAggregator with sub NestedAggregator may be not accurately
### Describe the bug

the result of NestedAggregator with sub NestedAggregator is not accurately here, the two values of doc_count should be 4.
image

### Related component

Search:Aggregations

### To Reproduce

1. create the index.
```
PUT index1_nest111
{
    "settings": {
    "index.refresh_interval":"30s"
    }, 
   "mappings": {
      "properties": {
         "nested1": {
            "type": "nested",
            "properties": {
               "name": {
                  "type": "keyword"
               }
            }
         },
         "nested2": {
            "type": "nested",
            "properties": {
               "age": {
                  "type": "long"
               }
            }
         }
      }
   }
}
```

2. put the data.
the 4 documents are same, except for the _id:
```
POST _bulk?refresh=true
{ "index": { "_index": "index1_nest111", "_id": "1" } }
{ "nested2": {"age":1}, "nested1": {"name": "name1"} }
{ "index": { "_index": "index1_nest111", "_id": "2" } }
{ "nested2": {"age":1}, "nested1": {"name": "name1"} }


POST _bulk?refresh=true
{ "index": { "_index": "index1_nest111", "_id": "3" } }
{ "nested2": {"age":1}, "nested1": {"name": "name1"} }
{ "index": { "_index": "index1_nest111", "_id": "4" } }
{ "nested2": {"age":1}, "nested1": {"name": "name1"} }
```

3. aggregation
```
POST index1_nest111/_search
{
  "aggregations": {
    "out_nested": {
      "aggregations": {
        "out_terms": {
          "aggregations": {
            "inner_nested": {
              "aggregations": {
                "inner_terms": {
                  "terms": {
                    "field": "nested1.name"
                  }
                }
              },
              "nested": {
                "path": "nested1"
              }
            }
          },
          "terms": {
            "field": "nested2.age"
          }
        }
      },
      "nested": {
        "path": "nested2"
      }
    }
  },
  "size": 0
}
```

### Expected behavior

The inner_nested.doc_count shouble alse be 4.

If it's a bug, I'm please to fix.

### Additional Details

**Host/Environment (please complete the following information):**
 - OS: os2.9

### PR DESCRIPTION
**Pull request by kkewwei**
Signed-off-by: kkewwei [kkewwei@163.com](mailto:kkewwei@163.com)

### Description
 The  aggs result of NestedAggregator with sub NestedAggregator may be not accurately, it depends on the face that the parentDocId > childDocId in search, but it may be wrong.

### Related Issues
Resolves #13303

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jed326** [comment]
@kkewwei thanks for opening the PR, it looks like the existing nested agg integration tests are failing after your change, could you please take a look? https://build.ci.opensearch.org/job/gradle-check/37514/#showFailuresLink

Also would you mind adding a changelog item as well? Thanks!

**Comment by kkewwei** [comment]
@jed326 I will process soon.

**Comment by kkewwei** [comment]
@jed326  please help review.

**Comment by kkewwei** [comment]
> UNSTABLE

testThreeNodesNoClusterManagerBlock  #10006

**Comment by kkewwei** [comment]
> ❕ Gradle check result for [a95ca76](https://github.com/opensearch-project/OpenSearch/commit/a95ca762fe0b796daf6ed0c7cfe4361f5c06fd3f): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/37808/)
> 
> * **TEST FAILURES:**
> 
> ```
>       3 org.opensearch.cluster.MinimumClusterManagerNodesIT.testThreeNodesNoClusterManagerBlock
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

testThreeNodesNoClusterManagerBlock #10006

**Comment by jed326** [comment]
Thanks @kkewwei, could you add an entry to [`CHANGELOG.md`](https://github.com/opensearch-project/OpenSearch/blob/main/CHANGELOG.md) for this as well?

**Comment by jed326** [comment]
Other than that I think this PR looks good to me, @msfroh @reta mind giving a maintainer review here as well?

**Comment by kkewwei** [comment]
> ❌ Gradle check result for [f3e4c6f](https://github.com/opensearch-project/OpenSearch/commit/f3e4c6f83694042c5d7c27566ff2be1f97ef3254): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/38033/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

testReadRangeBlobWithRetries  #6090
testIndexingWithReplicaOnBwcNodes #9407
testSeqNoCheckpoints #9407
testIndexingWithPrimaryOnBwcNodes  #9407
testIndexVersionPropagation #9407
MixedClusterClientYamlTestSuiteIT #2440

**Comment by kkewwei** [comment]
> testReadRangeBlobWithRetries #6090
> testIndexingWithReplicaOnBwcNodes #9407
> testSeqNoCheckpoints #9407
> testIndexingWithPrimaryOnBwcNodes #9407
> testIndexVersionPropagation #9407
> MixedClusterClientYamlTestSuiteIT #2440

testReadRangeBlobWithRetries https://github.com/opensearch-project/OpenSearch/issues/6090
testIndexingWithReplicaOnBwcNodes https://github.com/opensearch-project/OpenSearch/issues/9407
testSeqNoCheckpoints https://github.com/opensearch-project/OpenSearch/issues/9407
testIndexingWithPrimaryOnBwcNodes https://github.com/opensearch-project/OpenSearch/issues/9407
testIndexVersionPropagation https://github.com/opensearch-project/OpenSearch/issues/9407
MixedClusterClientYamlTestSuiteIT https://github.com/opensearch-project/OpenSearch/issues/2440

**Comment by kkewwei** [comment]
> ❌ Gradle check result for [6a222a8](https://github.com/opensearch-project/OpenSearch/commit/6a222a8d95d8653ea161980b6d6dc7c439d11d60): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/38038/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

testReadRangeBlobWithRetries https://github.com/opensearch-project/OpenSearch/issues/6090
testIndexingWithReplicaOnBwcNodes https://github.com/opensearch-project/OpenSearch/issues/9407
testSeqNoCheckpoints https://github.com/opensearch-project/OpenSearch/issues/9407
testIndexingWithPrimaryOnBwcNodes https://github.com/opensearch-project/OpenSearch/issues/9407
testIndexVersionPropagation https://github.com/opensearch-project/OpenSearch/issues/9407
MixedClusterClientYamlTestSuiteIT https://github.com/opensearch-project/OpenSearch/issues/2440

File: CHANGELOG.md
Comment (jed326) [inline CHANGELOG.md:90]:
```suggestion
- Fix aggregation result of NestedAggregator when it has a subAggregation with another NestedAggregator ([#13324](https://github.com/opensearch-project/OpenSearch/pull/13324))
```

**Comment by jed326** [comment]
Thanks @kkewwei , could you try to rebase from `main` to see if that fixes the flaky tests? We need to get the checks to green to be able to merge.

File: CHANGELOG.md
Comment (jed326) [reply CHANGELOG.md:90]: Changelog format spec: https://keepachangelog.com/en/1.0.0/

**Comment by dblock** [comment]
Can @msfroh or @andrross take a look please? Or volunteer another maintainer?

**Review by jed326 — APPROVED** [review]

File: server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:126]: @kkewwei just to clarify 

> // if parentDoc is 0 then this means that this parent doesn't have child docs (b/c these appear always before the parent

this is not valid comment?

**Comment by reta** [comment]
> @reta @msfroh @andrross would one of you be able to give a maintainer review here?

@msfroh I think we need your help, I have difficulties understanding the impact of this change (sorry @jed326 @kkewwei )

File: server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:126]: I need to dig into this. 

The parent/child block join in Lucene relies on the behavior of the old comment -- child documents are written immediately before their parent.

File: server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java
Comment (msfroh) [inline server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:149]: @kkewwei, correct me if I'm wrong, but this branch would only get reached if:
1. `parentDoc` already points to the first parent doc, in which case `currentParentDoc == parentDoc`.
2. `parentDoc` points to a child of the first parent, in which case `currentParentDoc > parentDoc`.

If `parentDoc` *doesn't* point to a parent document, I believe there's a mistake somewhere else in the code.

File: server/src/test/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregatorTests.java
Comment (msfroh) [inline server/src/test/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregatorTests.java:1067]: This result looks wrong:

1. There is no doc with ID 1. Lucene doc ids do not have gaps.
2. The child document with ID 2 must have been indexed as a child of parent with ID 3. This return value suggests that the child with ID 2 is associated with the parent with ID 0.

File: server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java
Comment (kkewwei) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:126]: I aggree that the child documents are written immediately before their parent. the `parentDoc` here is not the documnet parent, it should be more appropriately called the `searchParentDoc`. 

For example, nested2 can be `parentDoc`, but it is not the documnet parent.
```
POST index1_nest111/_search
{
  "aggregations": {
    "out_nested": {
      "aggregations": {
        "out_terms": {
          "aggregations": {
            "inner_nested": {
              "aggregations": {
                "inner_terms": {
                  "terms": {
                    "field": "nested1.name"
                  }
                }
              },
              "nested": {
                "path": "nested1"
              }
            }
          },
          "terms": {
            "field": "nested2.age"
          }
        }
      },
      "nested": {
        "path": "nested2"
      }
    }
  },
  "size": 0
}
```

Comment (kkewwei) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:126]: https://private-user-images.githubusercontent.com/16730433/324480985-7401b8a9-2648-4656-bc53-b730107fab24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUxNjAyNjgsIm5iZiI6MTcxNTE1OTk2OCwicGF0aCI6Ii8xNjczMDQzMy8zMjQ0ODA5ODUtNzQwMWI4YTktMjY0OC00NjU2LWJjNTMtYjczMDEwN2ZhYjI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA4VDA5MTkyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyMjE3ZWFhNjhiZmM2M2ExMjdkMGY5OGYyYjg5OTZkMzYwOTkyYzUxMTk3NjJmNzY5ZDYxOWY5YWFjYTI5OGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NB-wYtVAodrBBO4_WO-mgCqiScWsOj4zacrNQyQqxBk

Comment (jed326) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:126]: @msfroh The issue is specifically for the nested NestedAggregator subagg (confusing!)

@kkewwei is creating 2 segments in his reproduction https://github.com/opensearch-project/OpenSearch/issues/13303#issuecomment-2069477578 and we are basically missing the first document in the sub-agg for each segment.

File: server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java
Comment (kkewwei) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:149]: @msfroh, the two points are right.  (when aggregation, nest2 is outer path, nest1 is inner path)
first case:
image
parentDoc can be 1,3,5

second case:
image
parentDoc can be 0,3,6

Comment (kkewwei) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:149]: In the first case, field `nest2` is the parent doc of `nest1`(in which case currentParentDoc == parentDoc)
In the second case, field `nest2` and field `nest1` is independant,  `nest2` is behind/after`nest1 `.

File: server/src/test/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregatorTests.java
Comment (kkewwei) [reply server/src/test/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregatorTests.java:1067]: @msfroh Yes, you are right.
1. I will fix the logical after you finish reviewing the all code.
2. at the end of `getChildAndParentId`, we should add a check:  if currentParentDoc<=childDocId, childDocId = DocIdSetIterator.NO_MORE_DOCS.

please see other code.

**Comment by kkewwei** [comment]
> FAILURE

AwarenessAttributeDecommissionIT.testConcurrentDecommissionAction #12197
IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode @13600

**Comment by kkewwei** [comment]
@msfroh please see again.

**Comment by jed326** [comment]
@msfroh could you take another look at this one? Thanks

**Comment by kkewwei** [comment]
@msfroh, please have a look when you are free, or can you recommend someone else who has free time to take a look?

File: server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:128]: Ahh... so the problem come down to a mismatch on the word "parent". 

In one case, it's talking about the parent doc (from the Lucene block-join standpoint), while in the other case, it's talking about a child doc ID (from the block-join standpoint) from the parent aggregation, where we're trying to aggregate over a sibling of that child. So, we need to map from that document to its parent, then join to the appropriate sibling. 

Is that a fair representation of the situation?

File: server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:151]: Assuming my other comment is correct, it would probably be a good idea to add some comments here and rename some variables to explain the situation. 

Maybe the `parentDoc` parameter should be called `parentAggDoc`. If it's a top-level aggregation, it's reasonable that it would be the same as the block's parent doc. If it's also a `nested` aggregation, then it may be a sibling (in the block).

So, while the input parameter passed in here is the parent aggregation's current doc (which may or may not be a block-level parent doc), what we output  is a tuple consisting of the current block-level parent doc (the parent of the parameter doc), and the next matching child doc (hopefully under this parent) for the aggregation (according to the child doc iterator). 

If the next matching child doc is *not* under the block-parent of the current agg-parent's doc, then we return `NO_MORE_DOCS` as a signal that we have no matches for the current parent agg bucket. Actually, I think that part is unnecessary because of the condition on the for-loop on line 137. So, I think you can remove the part from line 168-170.

@kkewwei -- have I understood correctly?

**Comment by msfroh** [comment]
Sorry for the delay -- I think I finally understand what's going on here. The word "parent" seems to be overloaded, as it considers both the block-join parent (what I was assuming previously) and the parent aggregation (which may be collecting block-join parents, but might also be collecting block-join children).

I replied to some of my earlier comments -- I think we need some very clear code comments to explain the situation and clarify when the word "parent" means the parent aggregation versus when it means the block-join parent.

Thanks for your patience, and sorry for the confusion, @kkewwei!

**Comment by kkewwei** [comment]
@msfroh I will deal with as soon as possible.

File: server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java
Comment (kkewwei) [reply server/src/main/java/org/opensearch/search/aggregations/bucket/nested/NestedAggregator.java:128]: Yes, that's what I mean, I should distinguish it from the name.

**Review by msfroh — APPROVED** [review]

**Review by andrross — APPROVED** [review]

**Comment by andrross** [comment]
I just rebased and fixed the changelog conflict. Should be good to merge once tests pass again.

**Comment by kkewwei** [comment]
@andrross please help merge.

**Comment by jed326** [comment]
There was another CHANGELOG conflict, pushed a resolution

**Review by andrross — APPROVED** [review]

**Comment by andrross** [comment]
Thanks @kkewwei! I promise we'll get this merged :)

**Review by andrross — APPROVED** [review]

**Pull request merged**
